File: backend-ci/gestion/services/user_service.py
from django.db import transaction
from django.shortcuts import get_object_or_404
from gestion.models.user import User
from gestion.enums import Role


class UserService:

    @staticmethod
    @transaction.atomic
    def create(data):
        """Créer un nouvel utilisateur"""
        user = User.objects.create_user(
            username=data['email'],
            email=data['email'],
            password=data['password'],
            prenom=data['prenom'],
            nom=data['nom'],
            telephone=data.get('telephone', ''),
            role=data.get('role', Role.PELERIN)
        )
        return user

    @staticmethod
    def get_all():
        """Récupérer tous les utilisateurs"""
        return User.objects.all()

    @staticmethod
    def get_by_id(user_id):
        """Récupérer un utilisateur par son ID"""
        return get_object_or_404(User, id=user_id)

    @staticmethod
    def get_by_email(email):
        """Récupérer un utilisateur par son email"""
        return get_object_or_404(User, email=email)

    @staticmethod
    def update(user, data):
        """Mettre à jour un utilisateur"""
        for key, value in data.items():
            setattr(user, key, value)
        user.save()
        return user

    @staticmethod
    def delete(user):
        """Supprimer un utilisateur"""
        user.delete()

    @staticmethod
    def get_pelerins():
        """Récupérer tous les pèlerins"""
        return User.objects.filter(role=Role.PELERIN)

    @staticmethod
    def get_admins():
        """Récupérer tous les admins"""
        return User.objects.filter(role=Role.ADMIN)

File: frontend-ci/src/services/user.service.ts
/**
 * Service pour les utilisateurs
 * Correspond à gestion/services/user_service.py
 */

import apiClient from './api';
import { User, SaveUserDTO, UpdateUserDTO, CreateResponse } from '../types';

const ENDPOINTS = {
  BASE: '/users/',
  DETAIL: (id: number) => `/users/${id}/`,
};

export const userService = {
  /**
   * Récupérer tous les utilisateurs
   */
  async getAll(): Promise<User[]> {
    const response = await apiClient.get<User[]>(ENDPOINTS.BASE);
    return response.data;
  },

  /**
   * Récupérer un utilisateur par son ID
   */
  async getById(id: number): Promise<User> {
    const response = await apiClient.get<User>(ENDPOINTS.DETAIL(id));
    return response.data;
  },

  /**
   * Créer un nouvel utilisateur
   */
  async create(data: SaveUserDTO): Promise<CreateResponse> {
    const response = await apiClient.post<CreateResponse>(ENDPOINTS.BASE, data);
    return response.data;
  },

  /**
   * Mettre à jour un utilisateur
   */
  async update(id: number, data: UpdateUserDTO): Promise<User> {
    const response = await apiClient.put<User>(ENDPOINTS.DETAIL(id), data);
    return response.data;
  },

  /**
   * Supprimer un utilisateur
   */
  async delete(id: number): Promise<void> {
    await apiClient.delete(ENDPOINTS.DETAIL(id));
  },
};

export default userService;
